pkg/config/apiserver.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+// Package config APIServer defines and implements the config API.
+// The server constructs and uses a validator for validations
+// The server uses KVStore to persists keys.
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	restful "github.com/emicklei/go-restful"
+	"github.com/golang/glog"
+	rpc "github.com/googleapis/googleapis/google/rpc"
+
+	"istio.io/mixer/pkg/adapter"
+	pb "istio.io/mixer/pkg/config/proto"
+	"istio.io/mixer/pkg/expr"
+	"istio.io/mixer/pkg/status"
+)
+
+type validateFunc func(cfg map[string]string) (rt *Validated, ce *adapter.ConfigErrors)
+
+type readBodyFunc func(r io.Reader) ([]byte, error)
+
+// API is the server wrapper that listens for incoming requests to the manager and processes them
+type API struct {
+	version  string
+	rootPath string
+
+	// used at the back end for validation and storage
+	store    KVStore
+	validate validateFunc
+
+	// house keeping
+	handler http.Handler
+	Server  *http.Server
+
+	// fault injection
+	readBody readBodyFunc
+}
+
+// DefaultAPIPort default port exposed by the API server.
+const DefaultAPIPort = 9094
+
+// register routes
+func (a *API) register(c *restful.Container) {
+	ws := &restful.WebService{}
+	ws.Consumes(restful.MIME_JSON, "application/yaml", "application/x-yaml")
```

Comment:
YAML is probably best left to user-facing client-side only. There's a lot of complexity in YAMLs (imports, self-references) that need not be in the server-side.